### PR TITLE
Collection rarity based on normalised information content

### PR DESCRIPTION
**Pull request by aschlosberg**
All rationale is explained in the documentation of `rarity.go`.

- - -
This change is [Reviewable](https://reviewable.io/reviews/divergencetech/ethier/51)

### CODE REVIEW

File: erc721/erc721.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	Image        string       `json:"image"`
+	AnimationURL string       `json:"animation_url,omitempty"`
+	ExternalURL  string       `json:"external_url,omitempty"`
+	Attributes   []*Attribute `json:"attributes,omitempty"`
+}
+
+// An Attribute is a single attribute in Metadata.
+type Attribute struct {
+	TraitType   string      `json:"trait_type,omitempty"`
+	Value       interface{} `json:"value"`
```
Comment (cxkoda) [inline erc721/erc721.go:35]: I read somewhere that `any` is more common now. Is that true?

File: erc721/erc721.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	DisplayType string      `json:"display_type,omitempty"`
```
Comment (cxkoda) [inline erc721/erc721.go:36]: Since we're already following the OS-defined standard, should we use an enum for this?

File: erc721/rarity.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+// of the Collection as a whole (see the Rarity struct for rationale).
+//
+// Notably, the lack of a TraitType is considered as a null-Value Attribute as
+// the absence across the majority of a Collection implies rarity in those
+// tokens that do carry the TraitType.
+//
+// Non-string Attribute Values are passed to the bucket function. The returned
+// bucket is used in place of original value. It is valid for the bucket
+// function to simply return the string equivalent (e.g. true/false for
+// booleans).
```
Comment (cxkoda) [inline erc721/rarity.go:42]: why is this called bucket?

File: erc721/rarity.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+func (coll Collection) Rarity(bucket func(interface{}) string) *Rarity {
+
+	// distribution and counts carry floats instead of integers to make
+	// calculation of entropy simpler. counts[x] will contain the sum of all
+	// values in distribution[x], which is split by value.
+	distributions := make(map[string]map[string]float64)
+	counts := make(map[string]float64)
+	attributes := make(map[TokenID]map[string]string)
```
Comment (cxkoda) [inline erc721/rarity.go:50]: should we make the types more verbose?
```suggestion
        type AttributeType string
        type AttributeValue string
	distributions := make(map[AttributeType]map[AttributeValue]float64)
	counts := make(map[AttributeType]float64)
	attributes := make(map[TokenID]map[AttributeType]AttributeValue)
```

File: erc721/rarity.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+		attributes[id] = make(map[string]string)
+		for _, attr := range meta.Attributes {
+			counts[attr.TraitType]++
+
+			val, ok := attr.Value.(string)
+			if !ok {
+				// We can't make any assumptions about non-string discrete types
+				// (e.g. booleans) because they can technically clash with
+				// equivalent string values. Lesson: strong typing is important
+				// and JSON is stupid.
```
Comment (cxkoda) [inline erc721/rarity.go:62]: :+1:

File: erc721/rarity.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+				n = collSize - numHaveAttr
+			}
+
+			scores[id] += -math.Log2(n / collSize)
+		}
+	}
+
+	// It's not valid to consider all decimal points so we limit the precision
+	// relative to the log of the collection size as this is what dictates
+	// precision of individual probabilities.
```
Comment (cxkoda) [inline erc721/rarity.go:106]: This is not completely correct IMO.

Intuitively I agree that `~1/N` determines the uncertainty in our probabilities and therefore the score (although I'd have to follow the details to be sure). but rounding the value is in any case less accurate (although negligibly so) than keeping the insignificant digits. So keeping the plain value with all digits is surely not invalid.
The rounding is more just for human convenience.

**Review by cxkoda — COMMENTED** [review]
Looks good to me. I double-checked the computation of entropy and self-information.

Should we add some tests?

**Review by cxkoda — CHANGES_REQUESTED** [review]
> Reviewed 3 of 4 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/divergencetech/ethier/51)* status: all files reviewed, 1 unresolved discussion (waiting on @aschlosberg)

___
*[`erc721/erc721.go` line 36 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N8cHRLiAYS7DafpXGcd-r2-36:-N90cUsp5QWXSMJdLlVe:b-yqbdts) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/erc721.go#L36)):*
<details><summary><i>Previously, aschlosberg (Arran Schlosberg) wrote…</i></summary><blockquote>

Making it an enum blocks any extended use cases for arbitrarily defined display types that someone may want to use.
</blockquote></details>

I agree with that. However, the whole structure of this package is very much tailored to the use with OS. So limiting this to OS supported values would reduce misuse.
___
*[`erc721/rarity.go` line 42 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N8cHRLiAYS7DafpXGce-r2-42:-N90d3U9EAG_CBovGgCT:bq657mq) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/rarity.go#L42)):*
<details><summary><i>Previously, aschlosberg (Arran Schlosberg) wrote…</i></summary><blockquote>

Borrowed from histogram "buckets" or "bins".
</blockquote></details>

ah I see, as in "give me the bin that you would like me to count this for".
I found this a little confusing initially - but I get it now.
___
*[`erc721/rarity.go` line 50 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N8cHRLiAYS7DafpXGce-r2-50:-N90eG97951s1wu3u4qm:b-pbow6w) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/rarity.go#L50)):*
<details><summary><i>Previously, aschlosberg (Arran Schlosberg) wrote…</i></summary><blockquote>

There's a trade-off between type safety and usability. Forcing type assertions from string to `type Foo string` is a huge pain so I don't think it's worth it for an internal type.
</blockquote></details>

I found it hard to understand which map key was for what when I read the code first. But I'm happy to keep it as it is.
___
*[`erc721/rarity.go` line 106 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N8cHRLiAYS7DafpXGce-r2-106:-N90eluNCQ2hRWF-qYxP:bazdshf) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/rarity.go#L106)):*
<details><summary><i>Previously, aschlosberg (Arran Schlosberg) wrote…</i></summary><blockquote>

I want to avoid people reading any meaning into the insignificant digits. Of course there's a [Wikipedia entry](https://en.wikipedia.org/wiki/Significant_figures#Arithmetic) about this. Sharing so we can discuss offline but have it recorded here.
</blockquote></details>

Had another think about this over night, but let's move it offline.

**Review by cxkoda — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/divergencetech/ethier/51)* status: all files reviewed, 2 unresolved discussions (waiting on @aschlosberg)

___
*[`erc721/rarity.go` line 113 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N90kKEoALICHd7e5b6m:-N90kKEoALICHd7e5b6n:bafrcz2) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/rarity.go#L113)):*
> ```Go
> 			return math.Round(f*pow) / pow
> 		}
> 	})()
> ```

Here a quick example. Assume we have a collection of 6 tokens with 1 trait. The trait has 3 different values A,B,C that occur 1,2,3 times.
The relative self-information score for tokens with A,B,C is `{1.77156, 1.08622, 0.685331}`. Rounding this with the above logic results in the values `{2., 1., 1.}`. So it would suggest that having B and C is as rare, which is not the case. 



_Code quote:_
```Go
	scale := (func() func(float64) float64 {
		precision := int(math.Floor(math.Log10(float64(len(coll)))))
		pow := math.Pow10(precision)
		return func(f float64) float64 {
			return math.Round(f*pow) / pow
		}
	})()
```

**Review by aschlosberg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/divergencetech/ethier/51)* status: 3 of 7 files reviewed, 2 unresolved discussions (waiting on @cxkoda)

___
*[`erc721/erc721.go` line 36 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N8cHRLiAYS7DafpXGcd-r2-36:-N9GbKKR4hc0CV6hFiCj:b-8n6uvh) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/erc721.go#L36)):*
<details><summary><i>Previously, cxkoda (David Huber) wrote…</i></summary><blockquote>

I agree with that. However, the whole structure of this package is very much tailored to the use with OS. So limiting this to OS supported values would reduce misuse.
</blockquote></details>

Done. This was such a pain!
___
*[`erc721/rarity.go` line 50 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N8cHRLiAYS7DafpXGce-r2-50:-N9GWIlGBvhfUmks1btK:b-lhuoxu) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/rarity.go#L50)):*
<details><summary><i>Previously, cxkoda (David Huber) wrote…</i></summary><blockquote>

I found it hard to understand which map key was for what when I read the code first. But I'm happy to keep it as it is.
</blockquote></details>

I've added type aliases to make this easier. They don't add type safety, but they make it more readable.
___
*[`erc721/rarity.go` line 106 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N8cHRLiAYS7DafpXGce-r2-106:-N9GWPeeECt7nbtdIz74:bg91ird) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/rarity.go#L106)):*
<details><summary><i>Previously, cxkoda (David Huber) wrote…</i></summary><blockquote>

Had another think about this over night, but let's move it offline.
</blockquote></details>

I've removed it after our discussion about the collection being the population not a sample and therefore without measurement error.
___
*[`erc721/rarity.go` line 113 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N90kKEoALICHd7e5b6m:-N9GWXER-wCCYrhpcVXV:b-896fix) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/rarity.go#L113)):*
<details><summary><i>Previously, cxkoda (David Huber) wrote…</i></summary><blockquote>

Here a quick example. Assume we have a collection of 6 tokens with 1 trait. The trait has 3 different values A,B,C that occur 1,2,3 times.
The relative self-information score for tokens with A,B,C is `{1.77156, 1.08622, 0.685331}`. Rounding this with the above logic results in the values `{2., 1., 1.}`. So it would suggest that having B and C is as rare, which is not the case. 
</blockquote></details>

Done.

**Review by cxkoda — APPROVED** [review]
> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/divergencetech/ethier/51)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @aschlosberg)

___
*[`erc721/erc721.go` line 36 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N8cHRLiAYS7DafpXGcd-r2-36:-N9Gga47-zArjYAZ6AXY:bnxosu2) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/erc721.go#L36)):*
<details><summary><i>Previously, aschlosberg (Arran Schlosberg) wrote…</i></summary><blockquote>

Done. This was such a pain!
</blockquote></details>

Argh, sorry for that. Thanks!
___
*[`erc721/rarity.go` line 50 at r2](https://reviewable.io/reviews/divergencetech/ethier/51#-N8cHRLiAYS7DafpXGce-r2-50:-N9GgxXpD1stI7_b24fZ:b4pkhwa) ([raw file](https://github.com/divergencetech/ethier/blob/490ba82e6d164a168231979c5fe27f4e72655e55/erc721/rarity.go#L50)):*
<details><summary><i>Previously, aschlosberg (Arran Schlosberg) wrote…</i></summary><blockquote>

I've added type aliases to make this easier. They don't add type safety, but they make it more readable.
</blockquote></details>

Ohh I wasn't aware that there are type aliases in go. Thanks, this is great!

**Pull request merged**